app/models/user.rb

```diff
@@ -351,4 +274,150 @@ def pcode3
   def pcode4
     school.sierra_code
   end
+
+
+  # Another piece of code has determined this user is all set to be called
+  # complete and done.  Note: Usually, expect the next step to be the calling of
+  # Patron Service to create the user in Sierra.
+  def save_as_complete
+    self.status = STATUS_LABELS['complete']
+    self.save
+  end
+
+
+  # If the user's barcode is not yet finalized, then set its status to
+  # 'pending' and save.  In the future, there may be other conditions that
+  # could set the user to "pending", and we'll be checking for those here, as well.
+  def save_as_pending
+    # do we need to fill in a provisional barcode?
+    unless self.barcode.present?
+      self.barcode = self.assign_barcode!
+    end
+
+    self.status = STATUS_LABELS['barcode_pending']
+    self.save
+  end
+
+
+  # ################ THE BARCODES SECTION! ################
+
+  def assign_barcode!
+    LogWrapper.log('DEBUG', {
+       'message' => "Begin assigning barcode to #{self.email}",
+       'method' => "assign_barcode",
+       'status' => "start",
+       'user' => {email: self.email}
+      })
+
+    last_user_barcode = User.where('barcode < 27777099999999').order(:barcode).pluck(:barcode).last
+    # no non-nil barcodes found?  this should never happen, but let's make sure we can handle it
+    if last_user_barcode.blank?
+      last_user_barcode = 27777000000001
```

Comment:
This seems odd, to use a hardcoded barcode. Shouldn't we raise an exception? Or save and log an error... or...? If this is placeholder functionality, better to raise a `TODO` message exception for now.

Reply:
No, we've been allocated a specific barcode range for MLN.  The library lending branches give patrons plastic cards with a barcode, when the patrons register.  The cards are printed ahead of time, and transported to the branches in large-ish batches.  To prevent barcode conflicts, projects that create new barcodes outside of the branches are allocated ranges.  Our range is 27777000000001 to 27777099999999.  SimplyE has another range.  I'll put these numbers into app-level constants, so it's clear they're not mistakes, and can be set and used throughout.

Reply:
As to using the minimum as default, I'm assuming that this conditon can only arise if MLN doesn't yet have users.  Well, we do have some users outside of our barcode range, so I suppose I should check for that.  Anyway, I would expect this condition to never arise.